packages/host/tests/unit/ai-function-generation-test.ts

```diff
@@ -0,0 +1,318 @@
+    }
+    `);
+    let schema = cardApi.generatePatchCallSpecification(nestedCard);
+    assert.deepEqual(schema, {
+      type: 'object',
+      properties: {
+        thumbnailURL: { type: 'string' },
+        title: { type: 'string' },
+        description: { type: 'string' },
+        containerField: {
+          type: 'object',
+          properties: {
+            innerStringField: { type: 'array', items: { type: 'string' } },
+          },
+        },
+      },
+    });
+  });
+
+  test(`does not generate anything for linksTo`, async function (assert) {
+    let nestedCard = await createCard(`
+    import StringField from 'https://cardstack.com/base/string';
+    import {
+        Component,
+        CardDef,
+        FieldDef,
+        field,
+        linksTo,
+        contains
+    } from 'https://cardstack.com/base/card-api';
+
+    class OtherCard extends CardDef {
+        @field innerStringField = contains(StringField);
+    }
+
+    export class TestCard extends CardDef {
+        static displayName = 'TestCard';
+        @field linkedCard = linksTo(OtherCard);
+        @field simpleField = contains(StringField);
+    }
+    `);
+    let schema = cardApi.generatePatchCallSpecification(nestedCard);
+    assert.deepEqual(schema, {
+      type: 'object',
+      properties: {
+        thumbnailURL: { type: 'string' },
+        title: { type: 'string' },
+        description: { type: 'string' },
+        simpleField: { type: 'string' },
+      },
+    });
+  });
+
+  test(`skips over fields that can't be recognised`, async function (assert) {
+    let nestedCard = await createCard(`
+    import StringField from 'https://cardstack.com/base/string';
+    import {
+        Component,
+        CardDef,
+        FieldDef,
+        field,
+        contains,
+        primitive
+    } from 'https://cardstack.com/base/card-api';
+
+    class NewField extends FieldDef {
+        static displayName = 'NewField';
+        static [primitive]: number;
+    }
+
+    export class TestCard extends CardDef {
+        static displayName = 'TestCard';
+        @field keepField = contains(StringField);
+        @field skipField = contains(NewField);
+    }
+    `);
+    let schema = cardApi.generatePatchCallSpecification(nestedCard);
+    assert.deepEqual(schema, {
+      type: 'object',
+      properties: {
+        thumbnailURL: { type: 'string' },
+        title: { type: 'string' },
+        description: { type: 'string' },
+        keepField: { type: 'string' },
+      },
+    });
+  });
+
+  test(`Doesn't break when there's a loop`, async function (assert) {
+    let nestedCard = await createCard(`
+    import StringField from 'https://cardstack.com/base/string';
+    import {
+        Component,
+        CardDef,
+        FieldDef,
+        field,
+        contains,
+    } from 'https://cardstack.com/base/card-api';
+
+    class RecursiveField extends FieldDef {
+        @field innerRecursiveField = contains(() => RecursiveField);
```

Comment:
this is actually impossible--this won't type check. but you can't see that because you are using strings instead of code to make cards. `conains` can never include the enclosing card.